TeXmacs/progs/convert/markdown/markdownout.scm

```diff
@@ -0,0 +1,483 @@
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;
+;; MODULE      : tmmarkdown.scm
+;; DESCRIPTION : markdown-stree to markdown-document or markdown-snippet
+;; COPYRIGHT   : (C) 2017 Ana Cañizares García and Miguel de Benito Delgado
+;;
+;; This software falls under the GNU general public license version 3 or later.
+;; It comes WITHOUT ANY WARRANTY WHATSOEVER. For details, see the file LICENSE
+;; in the root directory or <http://www.gnu.org/licenses/gpl-3.0.html>.
+;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(texmacs-module (convert markdown markdownout)
+  (:use (convert tools output)))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; "Global" state for document serialization and config options
+;; Usage is wrapped within a "with-global" in serialize-markdown-document
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(define footnote-nr 0)
+(define label-nr 0)  ; global counter for labels (incl. labeled equations)
+(define equation-nr 0)
+(define environment-nr 0)  ; global counter for numbered environments
+(define num-line-breaks 2)
+(define paragraph-width #f)
+(define paper-authors '())
+(define post-tags '())
+(define post-author "")
+(define doc-title "")
+(define postlude "")
+(define labels '())
+(define indent "")
+(define (first-indent) indent)
+(define file? #f)
+
+(define (hugo-extensions?)
+  (== (get-preference "texmacs->markdown:hugo-extensions") "on"))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Helper routines
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; FIXME: does this make sense? We only convert if exporting to file.
+; The idea is that ""Copy to markdown" might already perform some internal
+; conversion before sending us the stree, because weird chars appear.
+; However if we don't do any conversion here, the copied text is still wrong
+(define (tm-encoding->md-encoding x)
+  (if file? (string-convert x "Cork" "UTF-8") x))
+
+(define (md-encoding->tm-encoding x)
+  (if file? (string-convert x "UTF-8" "Cork") x))
+
+(define (list->csv l)
+  (string-join (map (cut string-append "\"" <> "\"") l) ", "))
+
+(define (indent-increment s)
+  (string-append indent s))
+
+(define (author-add x)
```

Comment:
Note that this is not hugo specific, but a custom var for a blog of mine, so it probably doesn't make much sense to add it here. Instead we could use a preference with a list of custom texmacs tags and what they will be exported as in the markdown metadata. Right now there is (at least) `author-name` in the document being appended to `paper_authors` in the metadata. So the preference could look like a tuple with those two strings